### PR TITLE
Add pages-e2e

### PR DESCRIPTION
**Pull request by GregBrimble**
<details>
<summary>Beautiful inline errors</summary>

image
</details>

<details>
<summary>Simplified GitHub Actions workflow:</summary>

image
</details>

<details>
<summary>Comprehensive e2e test reporting</summary>

image
</details>

And I've pulled prettier to the monorepo so you don't need to keep redefining that everywhere.

Only problem here was two unit tests which mocked the console which I had to comment out: https://github.com/cloudflare/next-on-pages/pull/351/commits/84c1e8e655f9155eadacd08a8f48d5d74509a29c. Not sure what is going on with those.

### CODE REVIEW

**Comment by james-elicx** [comment]
Very exciting to get proper e2es 😄

I have a question about how the features work in fixtures. Are they built+run independently of one another? i.e., multiple conflicting features for a fixture won't interfere with each other running properly, or would we need different fixtures when features conflict? Essentially, is each fixture+feature combo built and tested separately?

On an unrelated note, I was looking at the [readme](https://www.npmjs.com/package/@cfpreview/pages-e2e-test-runner-cli?activeTab=readme) and it sounds like this depends on deploying to Pages to run the e2e tests. This might be a tad irritating when running them locally, especially if the internet isn't the best (me atm) or if one is offline. I imagine this is something that won't be a problem once local mode on the roadmap is done though.

p.s., since the test results are available remotely via a deployment in PRs, it would be kind of neat if it wrote a link to the test results as a comment on the PR, and updated it with each rerun, like with prereleases :)

**Review by dario-piotrowicz — COMMENTED** [review]
I've just reviewed the file changes and left some minor comments here and there 🙂 

I'm just dropping the comments now, I still need to play with this locally

**Comment by GregBrimble** [comment]
@james-elicx , the README is a tad out of date now (I'll update that once we've finalized exactly how we want it to work). But to address your comments:

> I have a question about how the features work in fixtures. Are they built+run independently of one another? i.e., multiple conflicting features for a fixture won't interfere with each other running properly, or would we need different fixtures when features conflict? Essentially, is each fixture+feature combo built and tested separately?

The test runner starts with each fixture. It looks at the `main.fixture` file and finds the array of `features`. It then creates a `workspace` for each one of the fixtures and applies the set of features to that workspace atop each other. So you have N fixtures and M features, you get N workspaces.

The tests from each of the fixtures and features are also combined, so if the fixture has a test and you're applying features which each have a test, you'd have three tests in that workspace.

Finally, you can also have global tests (`pages-e2e/__tests__`) which are also brought into every workspace as well.

That's the state of the world today. However, Dario has also asked for (and we have an internal need for): variants. These would allow you to run each fixture with some subtle changes. The bones of the fixture would stay the same, but you'd be able to modify a package's version or whatever. This hasn't yet been built, but it's coming soon.

> On an unrelated note, I was looking at the [readme](https://www.npmjs.com/package/@cfpreview/pages-e2e-test-runner-cli?activeTab=readme) and it sounds like this depends on deploying to Pages to run the e2e tests. This might be a tad irritating when running them locally, especially if the internet isn't the best (me atm) or if one is offline. I imagine this is something that won't be a problem once local mode on the roadmap is done though.

These tests are configured to run with [`--environment=local`](https://github.com/cloudflare/next-on-pages/pull/351/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R18) which means it uses Wrangler 3 to spin up a "deployment" rather than `--environment=production` or `--environment=staging` which spins up real deployments on Cloudflare.

> p.s., since the test results are available remotely via a deployment in PRs, it would be kind of neat if it wrote a link to the test results as a comment on the PR, and updated it with each rerun, like with prereleases :)

Totally could, but it does also appear already in the "test summary" of the Action's run.

image

Example: https://github.com/cloudflare/next-on-pages/actions/runs/5422274945?pr=351

File: package.json [hunk 1]
```diff
@@ -3,20 +3,31 @@
...
+		"test:e2e": "npx @cfpreview/pages-e2e-test-runner-cli --environment=local",
 		"publish": "turbo build & changeset publish",
 		"changeset": "npx changeset"
 	},
 	"devDependencies": {
+		"@cfpreview/pages-e2e-test-runner-cli": "^0.0.17",
+		"prettier": "^2.8.8",
 		"turbo": "^1.10.3"
+	},
+	"prettier": {
```
[the remark this answers is not in the document]
Comment (GregBrimble) [reply package.json:27]: This looks to be on by default :)

File: turbo.json [hunk 2]
```diff
@@ -10,7 +10,6 @@
 			"outputs": ["dist/**"]
 		},
 		"lint": {},
-		"prettier:check": {},
```
[the remark this answers is not in the document]
Comment (GregBrimble) [reply turbo.json:13]: Follow up in another PR to re-turbo-ify

File: .gitignore [hunk 3]
```diff
@@ -178,3 +178,7 @@ dist
 
 # Turobrepo
 .turbo/
+
+# pages-e2e
+/pages-e2e/workspaces
+/pages-e2e/test-workspaces
```
[the remark this answers is not in the document]
Comment (GregBrimble) [reply .gitignore:184]: Will follow up with a PR to the CLI which cleans by default and offers a new `--persist-workspaces` arg or similar.

**Review by dario-piotrowicz — APPROVED** [review]

**Comment by james-elicx** [comment]
> > I have a question about how the features work in fixtures. Are they built+run independently of one another? i.e., multiple conflicting features for a fixture won't interfere with each other running properly, or would we need different fixtures when features conflict? Essentially, is each fixture+feature combo built and tested separately?
> 
> The test runner starts with each fixture. It looks at the `main.fixture` file and finds the array of `features`. It then creates a `workspace` for each one of the fixtures and applies the set of features to that workspace atop each other. So you have N fixtures and M features, you get N workspaces.
> 
> The tests from each of the fixtures and features are also combined, so if the fixture has a test and you're applying features which each have a test, you'd have three tests in that workspace.
> 
> Finally, you can also have global tests (`pages-e2e/__tests__`) which are also brought into every workspace as well.
> 
> That's the state of the world today. However, Dario has also asked for (and we have an internal need for): variants. These would allow you to run each fixture with some subtle changes. The bones of the fixture would stay the same, but you'd be able to modify a package's version or whatever. This hasn't yet been built, but it's coming soon.

Thank you very much for expanding on this part - it definitely clears things up for me with how they work.

**Pull request merged**